gc/src/gc.rs

```diff
@@ -3,11 +3,17 @@ use std::mem;
 use std::cell::{Cell, RefCell};
 use trace::Trace;
 
-// XXX Obviously not 100 bytes GC threshold - choose a number
-const GC_THRESHOLD: usize = 100;
+const INITIAL_THRESHOLD: usize = 100;
+
+// after collection we want the the ratio of used/total to be no
+// greater than this (the threshold grows exponentially, to avoid
+// quadratic behavior when the heap is growing linearly with the
+// number of `new` calls):
+const USED_SPACE_RATIO: f64 = 0.7;
```

Comment:
That would probably be nicer. Before this patch it was literally just an arbitrary number which was lifted into a constant.

If you want to make a PR which makes it an environment variable, I'd probably merge it (with @Manishearth's approval ofc)